### ISSUE TITLE
Accessibility requirement: capacity to have an empty alt output in the HTML from an empty or null string variable passed into a th:alt property

### ISSUE THREAD

**Issue by Sven-WEBDESIGN** [thymeleaf/thymeleaf#765]
* Nature of the issue according to me: BUG
  * Problem description: Web accessibility for images that are decorative (that is most images on the web) requires that they have an empty alt property, see: [https://www.w3.org/WAI/tutorials/images/decorative/](https://www.w3.org/WAI/tutorials/images/decorative/)
  * Work-around: duplicate img tag with two versions, one with an alt="" when variable is an empty string or null, the other with the th:alt="", when the variable is a non nempty, non null string
  * Version of Thymeleaf: 3
  * Environment: not relevant
  * Add an **empty** th:alt property to an image tag if variable passed in the th:alt property resolves to null or empty string without having to manage empty or null string and duplicating tag

HTML in thymeleaf, case: string altText is not empty ( set to 'test')
```
<th:block th:with=" altText = 'test' ">
    <img alt=""
        src="https://via.placeholder.com/90x32?text=NonEmptynonNullString"
        th:alt="${altText}"
        th:if="${not #strings.isEmpty(altText)}"
    >
    <img alt=""
        src="https://via.placeholder.com/90x32?text=emptyOrNullString"
        th:if="${altText == '' or altText == null}"
    >
</th:block>
```
Outputs
```
<img alt="" src="https://via.placeholder.com/90x32?text=NonEmptynonNullString">
```

HTML in thymeleaf, case: string altText is empty ( set to '')
```
<th:block th:with=" altText = '' ">
    <img alt=""
        src="https://via.placeholder.com/90x32?text=NonEmptynonNullString"
        th:alt="${altText}"
        th:if="${not #strings.isEmpty(altText)}"
    >
    <img alt=""
        src="https://via.placeholder.com/90x32?text=emptyOrNullString"
        th:if="${altText == '' or altText == null}"
    >
</th:block>
```
Outputs
```
<img alt="" src="https://via.placeholder.com/90x32?text=emptyOrNullString">
```

HTML in thymeleaf, case: string altText is empty ( set to null)
```
<th:block th:with=" altText = null ">
    <img alt=""
        src="https://via.placeholder.com/90x32?text=NonEmptynonNullString"
        th:alt="${altText}"
        th:if="${not #strings.isEmpty(altText)}"
    >
    <img alt=""
        src="https://via.placeholder.com/90x32?text=emptyOrNullString"
        th:if="${altText == '' or altText == null}"
    >
</th:block>
```
Outputs
```
<img alt="" src="https://via.placeholder.com/90x32?text=emptyOrNullString">
```

HTML in thymeleaf, case: string altText is empty (set to null or '')
It should be possible to output 
```
<th:block th:with=" altText = null ">
    <img
        src="https://via.placeholder.com/90x32?text=emptyOrNullString"
        th:alt="${altText}"
    >
</th:block>
<th:block th:with=" altText = '' ">
    <img
        src="https://via.placeholder.com/90x32?text=emptyOrNullString"
        th:alt="${altText}"
    >
</th:block>
```
Outputs
```
<img src="https://via.placeholder.com/90x32?text=emptyOrNullString">
```

Thymeleaf SHOULD output, for web accessibility needs
```
<img alt="" src="https://via.placeholder.com/90x32?text=emptyOrNullString">
```

**Comment by ultraq** [comment]
Hi there, I tried creating an `<img>` with an empty `alt` attribute, ran it through Thymeleaf, and the output was as expected: an empty `alt` attribute.

For this template:

```html
<body>
  <img alt="" src="https://via.placeholder.com/90x32/%20?Text=" >
</body>
```

This is what I got:

![Screen Shot 2019-10-11 at 10 52 59 AM](https://user-images.githubusercontent.com/1686920/66609358-4f468200-ec15-11e9-8b11-ad0e1223d1b1.png)

This was using Thymeleaf 3.0.11 w/ Spring Boot 2.1.4, though I expect the result to be the same even if Spring Boot wasn't in the mix.

I did however get errors if I tried to use `th:alt=""` - Thymeleaf could not parse an empty expression - so it might be worth checking if you ended-up with any empty `th:alt` attributes in your code.

**Comment by Sven-WEBDESIGN** [comment]
Hello, @ultrahq, you are right: i don't know how i messed up my test. I am thankful for you taking the time to test and have this feedback. I'll close the issue right now.

**Comment by Sven-WEBDESIGN** [comment]
I changed the case to refine the problem: there's no possibility to manage empty / null values in an th:alt property without having no outputted alt="" in the resulting HTML or having to duplicate blocks with an alt="" and a th:alt property with a non empty non null value.

Thymeleaf SHOULD output th:alt empty properties as alt="".

Thanks to @ultraq for helping me clarify the issue and pointing to the true nature of the problem (thymeleaf doesn't output empty value in th: properties).

**Comment by ultraq** [comment]
Yeah, the `th:alt=""` thing does bother me - I feel it should at least emit an empty `alt` value if the expression in it is empty itself or resolves to some "falsey" value.  I think we have other attributes that do this, so I can check those.

I'll leave this issue open and tag it as appropriate so that the other devs can see this and we can discuss if necessary.  Thanks for bringing this up @Sven-WEBDESIGN.

**Comment by Sven-WEBDESIGN** [comment]
Hello, thanks for considering this issue, it's an important point, accessibility-wise.
Let me know if i can help you in any way.

Have a good day.

**Comment by Sven-WEBDESIGN** [comment]
Hello, thanks for considering this issue, it's an important point, accessibility-wise.
Let me know if i cna help you in any way.

Have a good day.

**Comment by Sven-WEBDESIGN** [comment]
Sorry, i closed the issue by error.
Have a good day.

**Comment by Sven-WEBDESIGN** [comment]
Hello, would you be kind so kind to give me an idea when that feature will be taken into account so that i know if i have to implement an alternative in the meantime?
Thanks for that.

**Comment by ultraq** [comment]
Hi @Sven-WEBDESIGN, I'd suggest you go ahead and implement your alternative as we have no timeframe for this feature.

**Comment by Sven-WEBDESIGN** [comment]
Thanks for your feedback and its speed, i appreciate.
Have a good day.

**Comment by jbaiter** [comment]
@Sven-WEBDESIGN did you find a good workaround for this issue?

**Comment by anton-lilleby** [comment]
Hi, 
I had the samme issue and solved it client side @jbaiter 

```
const images = document.getElementsByTagName("img");
  for (let i = 0; i < images.length; i++) {
    if (!images[i].hasAttribute("alt")) {
      images[i].setAttribute("alt", "");
    }
  }
```

**Comment by xtianus** [comment]
I'd like to have `th:alt="${emptyString}"` output an empty alt, but maybe there should be a trick that doesn't break the standard behaviour of removing empty attributes. How about some kind of flag in the expression? It would be something like `th:alt="+${emptyString}"` where + means "keep the attribute when empty".
